crates/sui/src/unit_tests/cli_tests.rs

```diff
@@ -852,6 +854,140 @@ async fn test_package_publish_nonexistent_dependency() -> Result<(), anyhow::Err
     Ok(())
 }
 
+#[sim_test]
+async fn test_package_upgrade_command() -> Result<(), anyhow::Error> {
+    let mut test_cluster = TestClusterBuilder::new().build().await?;
+    let address = test_cluster.get_address_0();
+    let context = &mut test_cluster.wallet;
+
+    let client = context.get_client().await?;
+    let object_refs = client
+        .read_api()
+        .get_owned_objects(
+            address,
+            Some(SuiObjectResponseQuery::new_with_options(
+                SuiObjectDataOptions::new()
+                    .with_type()
+                    .with_owner()
+                    .with_previous_transaction(),
+            )),
+            None,
+            None,
+            None,
+        )
+        .await?
+        .data;
+
+    // Check log output contains all object ids.
+    let gas_obj_id = object_refs.first().unwrap().object().unwrap().object_id;
+
+    // Provide path to well formed package sources
+    let mut package_path = PathBuf::from(TEST_DATA_DIR);
+    package_path.push("dummy_modules_upgrade");
+    let build_config = BuildConfig::new_for_testing().config;
+    let resp = SuiClientCommands::Publish {
+        package_path: package_path.clone(),
+        build_config,
+        gas: Some(gas_obj_id),
+        gas_budget: 20_000,
+        skip_dependency_verification: false,
+        with_unpublished_dependencies: false,
+    }
+    .execute(context)
+    .await?;
+
+    // Print it out to CLI/logs
+    resp.print(true);
+
+    let SuiClientCommandResult::Publish(response) = resp else {
+        unreachable!("Invalid response");
+    };
+
+    let package = response
+        .effects
+        .as_ref()
+        .unwrap()
+        .created()
+        .iter()
+        .find(|refe| matches!(refe.owner, Owner::Immutable))
+        .unwrap();
+
+    let cap = response
+        .effects
+        .as_ref()
+        .unwrap()
+        .created()
+        .iter()
+        .find(|refe| matches!(refe.owner, Owner::AddressOwner(_)))
+        .unwrap();
+
+    // Hacky for now: we need to add the correct `published-at` field to the Move toml file.
+    // In the future once we have automated address management replace this logic!
+    let mut toml_path = package_path.clone();
+    toml_path.push("Move.toml");
+    let mut move_toml = std::fs::File::options()
+        .read(true)
+        .write(true)
+        .open(toml_path)
+        .unwrap();
+
+    let mut buf = String::new();
+    move_toml.read_to_string(&mut buf).unwrap();
+
+    // Add a `published-at = "0x<package_object_id>"` to the Move manifest.
+    let mut lines: Vec<String> = buf.split('\n').map(|x| x.to_string()).collect();
+    let idx = lines.iter().position(|s| s == "[package]").unwrap();
+    lines.insert(
+        idx + 1,
+        format!(
+            "published-at = \"{}\"",
+            package.reference.object_id.to_hex_uncompressed()
+        ),
+    );
+    let new = lines.join("\n");
+    move_toml.write_at(new.as_bytes(), 0).unwrap();
```

Comment:
This looks to be writing back to the file in the repository -- you're going to have a bad time with this in the presence of concurrent tests -- better to copy the whole package elsewhere (i.e. into a temporary directory) and then make your changes there.

Reply:
So I was going back-and-forth on this since this should be fine since this is the only test that is using that package. But, I shouldn't be lazy and do this properly. I'll update :)

Reply:
Should be all fixed now. Was much easier once I realized we were already using `fs_extra` so I could just use that to copy the directory over into a tempdir.